Image: Fix layout shift when lightbox is opened and closed

cbravobernal opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Currently, opening and closing the lightbox causes the website layout to shift, which can cause the zoom animation to appear broken, particularly on mobile. It is also possible to scroll the webpage while the lightbox is open by directly manipulating the scrollbar or using a mobile device, which is not in line with web standards.

This PR improves the lightbox implementation by removing the layout shift and updating the scroll handling so that the lightbox behaves reliably whenever a scroll is performed.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
We want to offer a polished lightbox experience, and the current experience appears broken in some cases. 

## How?
This PR removes `overflow: hidden` from the styles and updates the callback for the scroll handling.

Note: Expected behavior is that scrolling will be disabled on both desktop and mobile.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

1. Insert an image in a post and add enough content to ensure that the scrollbar will be visible.
2. Enable the lightbox using the block settings.
4. Publish and view the post on desktop.
5. Click on the image to open and close the lightbox — make sure that the scrollbar remains intact and that there is no layout shift.
6. Attempt scrolling while the lightbox is open using the mousewheel, `up` and `down` arrow keys, the `PageUp` and `PageDown` keys, and by directly clicking and dragging the scrollbar — verify that the lightbox stays open and doesn't scroll when performing any of those actions.
7. On mobile devices, test that scrolling is disabled and that the lightbox remains open when attempting to scroll using touch input. Test also that pinch to zoom while the lightbox is open works reasonably well.

## Screenshare

### Layout Shift Demonstration

#### Before the Fix

https://github.com/WordPress/gutenberg/assets/5360536/3789a30b-5df3-42d3-85aa-c4369f61bd77

#### After the Fix

https://github.com/WordPress/gutenberg/assets/5360536/3682b764-17a0-4964-9d9f-2f9cdc4f777f

michalczaplinski reviewed (approved):
Yeah, this seems reasonable to me 👍

cbravobernal commented:
I added the accesibility feedback label. Cause @artemiomorales mentioned there could be accesibility issues with this approach.

@alexstine can you help me here? Thanks!

alexstine commented:
CC: @joedolson or @afercia .

artemiomorales commented:
CC @andrewhayward

joedolson commented:
Allowing the background to scroll can be disorienting if a user dismisses the lightbox and is no longer in the same location on the page as they were previously. Focus is set on close, but that won't necessarily land the user in exactly the same place they were when they started.

What about using `overflow-y: hidden`, to avoid changing the scrollbar behavior?

There's no evidence that `has-lightbox-open` has any conflicts, surprisingly: https://wpdirectory.net/search/01H6ENN1605ZQ2B4Z7GZFZ61TW

cbravobernal commented:
> There's no evidence that has-lightbox-open has any conflicts, surprisingly: https://wpdirectory.net/search/01H6ENN1605ZQ2B4Z7GZFZ61TW

Renaming to wp-has-lightbox-open is not a problem, and we avoid future problems.

SantosGuillamot commented:
@joedolson  Just to clarify the use case and ensure I understand it correctly:
* When the user opens the lightbox it covers the whole screen, and there is no overflow. Neither in the X axis nor the Y axis.
* Scroll is not 100% disabled because, if the user scrolls vertically, it closes the lightbox and focuses on the previous element.

Should `overflow` be set to `hidden` in the HTML root, even if there is actually no overflow and the user can scroll to close the lightbox?

> Focus is set on close, but that won't necessarily land the user in exactly the same place they were when they started.

Regarding this, as part of [this pull request](https://github.com/WordPress/gutenberg/pull/52765), I [added the `preventScroll` property](https://github.com/WordPress/gutenberg/pull/52765/commits/a18c7213c0898f70159efb1188e9e670ddfe57be) to ensure the user lands in the same place where they started. Do you think that should be enough?

artemiomorales commented:
While testing, I noticed that the `overflow: hidden` property causes the lightbox zoom out animation to behave incorrectly in some cases.

I also found out that, if one scrolls quickly after opening the lightbox on iOS, that one is able to scroll even through the `overflow: hidden` property is enabled.

Since it seems like CSS may not be the best tool to prevent scrolling in this scenario considering that hiding the scrollbar causes a layout shift and potential errors with the lightbox calculations, I can explore how to prevent scrolling using JavaScript while the lightbox is open.

artemiomorales reviewed (changes requested):
Without `overflow: hidden`, one can scroll freely while the lightbox is open on iOS, breaking the experience.

We can either try to find another CSS solution or solve it with JavaScript. I'm leaning towards JavaScript because even with `overflow: hidden`, the experience breaks sometimes on mobile.

I'll look into addressing this as soon as I have bandwidth, but would be happy for someone else to hop onto this as well 😄

SantosGuillamot commented:
You're right that removing `overflow: hidden` allows the user to scroll, breaking the experience. Although I believe it is not an issue only in iOS:

https://github.com/WordPress/gutenberg/assets/34552881/0d7f0322-eb09-4718-9982-a9a5cb742f29

We don't see it on other devices because the lightbox is closed on `mousewheel`. Maybe we can solve it adding a listener to the `touchmove` to close it on mobile devices as well. Something like:

```
data-wp-on--wheel="actions.core.image.hideLightbox"
data-wp-on--touchmove="actions.core.image.hideLightbox"
```

By the way, it seems that **[mousewheel](https://developer.mozilla.org/en-US/docs/Web/API/Element/mousewheel_event) is deprecated**, so I guess we should use `wheel` instead.

Maybe another option is using [overscroll-behavior](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior), although I am not familiar with that.

Or just keep the `overflow: hidden`, but I'm aware it was causing some issues as well.

artemiomorales committed:
Replace overflow: hidden with JavaScript callback to prevent scrolling

artemiomorales commented:
@SantosGuillamot How does [this approach](https://github.com/WordPress/gutenberg/pull/53026/commits/1fa6b220c45f68a16a605dc28113dff33d97303a) look?

It feels a bit awkward to put the callback outside of the store, but I can't think of a good reason to have it otherwise.

Defining a general callback for the `scroll` event seems to handle the following scenarios:
- Scrolling with a mouse
- Scrolling via touch on mobile
- Scrolling via up and down arrows on keyboard
- Scrolling by dragging the scrollbar

I haven't tested on a physical mobile device yet so I'm not sure what the UX is like in that case, and dragging the scrollbar can feel a bit awkward, but otherwise it seems to work fairly well considering we're overriding the browser's native scrolling behavior.

SantosGuillamot commented:
> @SantosGuillamot How does [this approach](https://github.com/WordPress/gutenberg/pull/53026/commits/1fa6b220c45f68a16a605dc28113dff33d97303a) look?

I didn't have time to test it, and I'll be AFK for ten days. I am not a big fan of adding global `eventListeners`, but if we ensure that it only runs when the lightbox is open and only once, I guess it could work.

If I understood it correctly, the other option would be to use:

```
data-wp-on--wheel="actions.core.image.hideLightbox"
data-wp-on--touchmove="actions.core.image.hideLightbox"
```

And add the arrow keys in the Escape conditional: [link](https://github.com/WordPress/gutenberg/blob/1fa6b220c45f68a16a605dc28113dff33d97303a/packages/block-library/src/image/view-interactivity.js#L121).

And scrolling with the scrollbar wouldn't close the lightbox.

Is this right? I don't know which one is better, to be honest.

artemiomorales commented:
@SantosGuillamot The alternative approach you've stated would cover the following scenarios:

- Scrolling with a mouse
- Scrolling via touch on mobile
- Scrolling via up and down arrows on keyboard

However, it would not cover the following:

- Scrolling by dragging the scrollbar

In this last scenario, the lightbox would not close, and the user could also scroll while the lightbox is still open, which means that its origin reference image could be off, so the zoom animation would also break.

The `global` event listener seems to be the only way around this, and also handles the other scenarios.

artemiomorales committed:
Disable scroll callback on mobile; add comments; fix scrim styles

The page jumps around when trying to override the scroll behavior
on mobile, so I disabled it altogether. I've also added comments
to clarify this and other decisions made around the scroll handling.

While testing, I realized that the scrim was completely opaque during
the zoom animation, which does not match the design, so I added a new
animation specifically for the scrim to fix that.

artemiomorales commented:
@SantosGuillamot After testing on a physical device, the scroll handling, while it works, causes the page to jump around while swiping, so I've [disabled touch move input](https://github.com/WordPress/gutenberg/pull/53026/commits/a33564bf1693b62037cba726aa1d508b9d3f4e56) on mobile altogether. Mobile users can still close the lightbox by tapping, though not by scrolling, which I think seems reasonable.

artemiomorales committed:
Add handling for horizontally oriented pages

artemiomorales reviewed (approved):
I've now taken over this PR, so I could probably use another set of eyes on it, but these changes look good to me 😄

artemiomorales committed:
Move close button so that it's further from scrollbar on desktop

afercia commented:
Just noting that this PR lacks testing instructions, which I think are required for all pull requests. Regardless whether the testing flow may seem obvious, testing instructions must always be provided to allow all contributors, even the ones who may not be familiar with a specifici feature, to test and contribute.
WordPress is an open source project. We do want to encourage everyone to contribute. Please remind to always include clear test instructions.

afercia commented:
Not sure `overflow: hidden` should be removed, but I have a few questions:

1. Why the Lighbox closes when scrolling in the first place? It doesn't seem to me this is ideal and it's not an expected interaction. Users could inadvertently close the LIghtbox. I'm not sure implementing 'smart' behaviors is ideal, as they may be unexpected and even frustrating for some users. I'd just provide explicit actions to open / close.
2. If closing on scrolling stays, then it should work with any scrolling action. As far as I can see:
  - It closes when scrolling with mouse / trackpad.
  - It closes when scrolling with the Spacebar.
  - It does _not_ close when scrolling with Up and Down arrow keys.
  - It does _not_ close when scrolling with Page Up and Page Down arrow keys.
3. Why `overflow: hidden` is set to the `html` element? It should be set to the `body` element, as that's the actual scrollable element. Am I missing something?

artemiomorales committed:
Fix call to handleScroll() and add touch callback to new render method

artemiomorales commented:
@afercia Thank you for the feedback and sorry for the delay!

> Just noting that this PR lacks testing instructions, which I think are required for all pull requests. Regardless whether the testing flow may seem obvious, testing instructions must always be provided to allow all contributors, even the ones who may not be familiar with a specifici feature, to test and contribute.
WordPress is an open source project. We do want to encourage everyone to contribute. Please remind to always include clear test instructions.

Thank you for pointing this out. The PR description has been revised to include detailed test instructions.

> Why the Lighbox closes when scrolling in the first place? It doesn't seem to me this is ideal and it's not an expected interaction. Users could inadvertently close the LIghtbox. I'm not sure implementing 'smart' behaviors is ideal, as they may be unexpected and even frustrating for some users. I'd just provide explicit actions to open / close.

This is a common behavior on other sites and I'd argue that it allows users to better remain engaged with the content. I'll tag in @jasmussen and @jameskoster for additional thoughts.

> It does not close when scrolling with Up and Down arrow keys.
It does not close when scrolling with Page Up and Page Down arrow keys.

Thanks for catching this! That was a bug and it should now close in those circumstances. 

> Why overflow: hidden is set to the html element? It should be set to the body element, as that's the actual scrollable element. Am I missing something?

We've removed `overflow: hidden`, so I believe this no longer applies.

jasmussen commented:
Fundamentally, the lightbox is about providing a much simpler interface for quickly zooming an image in the content without breaking the reading flow by, for example, having to open an image in a new tab and zooming. Most of the behaviors on common lightbox implementations are geared towards that, including closing the lightbox on scroll, assuming the intention is to continue reading rather than continue looking at the picture.

I will trust you all to decide if this is possible to do in an accessible way. Question, however, and this is only speculation for a potential future enhancement: if it is not possible to do in an accessible way, would it be possible to add a toggle, off by default, that enabled "close on scroll"?

afercia commented:
> Most of the behaviors on common lightbox implementations are geared towards that, including closing the lightbox on scroll, assuming the intention is to continue reading rather than continue looking at the picture.

I'm skeptical on this being the only default behavior. I'm not sure it does matter whether this is a common pattern in other implementations, as it may may often that some patterns are just 'commonly wrong'. As you mentioned, it is a behavior based on an assumption. Assumptions aren't always the best option and I'd think we should leave users make a decision.

I'd like to hear thoughts other from a11 specialists, pinging @joedolson as this is mainly a visual thing. @alexstine although this is mainly visual, I think you can get an idea on what is going on anyways. Thanks.

Note:
While testing this, I noticed a few other functional and accessibility issues that I'll try to detail separately.

alexstine commented:
Thinking back to when I had low vision, I would hate it if something closed on scroll. Mainly because I used magnification software that often cut off what I was trying to look at so I would have to adjust the element in view to fit it all in. I would never recommend closing things on scroll. Not totally sure if I get the full point of this PR though so will allow others to weigh in.

joedolson commented:
Closing on scroll will also impact users with conditions that cause mouse instability, such as essential tremor, Parkinson's, Multiple Sclerosis, or any other conditions that result in instable control of environment. 

The image is already pretty easy to close; via Escape, clicking anywhere outside of the image, hitting the close button; all of those are already available. Closing on scroll just seems like an invitation to accidental closures. 

If, for example, you're using the Windows "Magnifier" tool, it enlarges your entire desktop so that only a portion is visible at any one time, with the locus of visibility centered on your mouse location. In order to see additional portions of the image - since the whole image won't be visible at once - you need to scroll to bring them into the viewable portion of the desktop.

andrewhayward commented:
> Most of the behaviors on common lightbox implementations are geared towards that, including closing the lightbox on scroll, assuming the intention is to continue reading rather than continue looking at the picture.

I'm certainly no expert on lightboxes, but I'll be honest, I'm not sure as I've ever found one that closes on scroll. Could you point out a few to experiment with?

> I will trust you all to decide if this is possible to do in an accessible way. Question, however, and this is only speculation for a potential future enhancement: if it is not possible to do in an accessible way, would it be possible to add a toggle, off by default, that enabled "close on scroll"?

Jut my opinion, but I'm not totally onboard with the idea that we would offer a setting for content creators knowing that it would be explicitly damaging for some content consumers. I think we have as much — if not more — responsibility to the latter group than we do the former.

jasmussen commented:
> I'm certainly no expert on lightboxes, but I'll be honest, I'm not sure as I've ever found one that closes on scroll. Could you point out a few to experiment with?

Sure, [this one](https://medium.com/niagara-launcher/introducing-the-widget-stack-7fef6c63501a). 

I hear you all. The reason I mention close on scroll is that it is a very common interaction especially on mobile devices. To an extent you can think of the TikTok scroll snapping as a modal interaction where the swiping behavior is a way out of that. 

> Jut my opinion, but I'm not totally onboard with the idea that we would offer a setting for content creators knowing that it would be explicitly damaging for some content consumers. I think we have as much — if not more — responsibility to the latter group than we do the former.

Understood.

I'd like to offer nuance here, that we offer a contrast warning when you choose a light yellow color on a white background, but we don't block you from doing it. We also offer a document outline tool to enable you to fix incorrect heading levels (and I'd like to see more here, surfacing incorrect heading levels in the pre-publish flow), but we don't block you from creating an incorrect hierarchy. In the future [it is also likely that it will be possible](https://github.com/WordPress/gutenberg/issues/48202#issuecomment-1442115759) to set a global alignment to enable justified text, as justified text is an (ill-advised) requirement in some newsrooms and even school classrooms. In that context, we would also offer a warning that it offers reduced legibility. Ultimately as a free-form editor, it will always be a spectrum. In this case, and because of the mobile use case, I offered the question of a toggle in good faith.

artemiomorales committed:
Improve lightbox experience on mobile

To ensure pinch to zoom works as expected when the lightbox
is open on mobile, I added logic to disable the scroll override
when touch is detected. Without this, the scroll override kicks
in whenever one tries to pinch to zoom, and it breaks the experience.

I also revised the styles for the scrim to make it opaque, as having
content visible outside of the lightbox is distracting when
pinching to zoom on a mobile device, and I think having a consistent
lightbox across devices will make for the best user experience.

artemiomorales committed:
Fix spacing

artemiomorales committed:
Add touch directives to block supports

artemiomorales committed:
Fix spacing in block supports

artemiomorales commented:
Thanks everyone for the discussion! I've gone ahead and disabled closing the lightbox on scroll 👍

I've also added a few improvements to make sure scrolling is indeed disabled when the lightbox is open on mobile. I've also added logic to ensure that pinch to zoom works as expected.

Worth noting is that, after testing, I've come to the conclusion that the scrim should be 100% opaque, as having any visible content in the background is distracting when one is doing the pinch to zoom action, and I think it's best to have a consistent user experience for the lightbox across devices.

I believe this PR is ready for a hopefully final review 😄

joedolson commented:
Making sure pinch to zoom works and a fully opaque scrim are both wins from my perspective. Thanks @artemiomorales!

dmsnell commented on packages/block-library/src/image/view.js:
```diff
@@ -17,6 +17,43 @@ const focusableSelectors = [
+// to is triggered on the window; so we define it outside of
+// the store to signal that the behavior here is different.
+// If we find a compelling reason to move it to the store,
+// feel free to do so.
```
if we turn these into JSDoc comments then the explanation will follow the variables and functions. otherwise they are only evident to those who look at this spot where they're created.

```js
/**
 * Stores a context-bound scroll handler.
 *
 * This callback could be defined inline inside of the store
 * object but it's created externally to avoid confusion about
 * it's logic…
 *
 * @var {function}
 */
let scrollCallback;

/** 
 * Tracks whether user is touching screen; used to differentiate mobile and desktop behaviors.
 *
 * @var {boolean}
 */
let isTouching = false;

/** 
 * Used to differentiate mobile and desktop behaviors.
 *
 * @var {boolean}
 */
let isTouching = false;

/**
 * Lightbox page-scroll handler: prevents scrolling.
 *
 * This handler is added to prevent scrolling behaviors that
 * trigger content shift while the lightbox is open…
 *
 * It would be better to accomplish this through CSS alone, but
 * because … a JavaScript version is required to avoid … in
 * order to provide the best visual experience.
 *
 * @param {object} context Interactivity page context?
 */
function handleScroll( context ) {
	…
}
```

artemiomorales replied:
Thanks for the heads up! I've converted these all to JSDoc comments.

dmsnell commented on packages/block-library/src/image/view.js:
```diff
@@ -17,6 +17,43 @@ const focusableSelectors = [
+		// because the scroll event can't be canceled, so we reset the position instead.
+		window.scrollTo(
+			context.core.image.lastScrollLeft,
+			context.core.image.lastScrollTop
```
I guess these are already defined and used, but they feel more like `context.core.image.initialScrollLeft` and `initialScrollTop` to me since they aren't updated; since they are resetting the scroll to the position it was in when initially opening the lightbox.

artemiomorales replied:
Got it! While `initialScrollTop` and `initialScrollLeft` may be better, they still don't ring true to me, so I've renamed them to `scrollTopReset` and `scrollLeftReset`, which I think more clearly expresses the intent.

dmsnell commented on packages/block-library/src/image/view.js:
```diff
@@ -43,54 +80,50 @@ store(
+							// animation is defined in the styles.scss and depends on if the
+							// animation is 'zoom' or 'fade', but in any case we should wait
+							// a few milliseconds longer than the duration, otherwise a user
+							// may scroll too soon and cause the animation to look sloppy.
```
is there a practical way to cancel the animation upon scroll without causing the page to shift? sorry if someone has already asked about this; it seems like another approach to avoid jitter in the page without actively preventing someone from doing what they're trying to do. asking because it can be irritating on a website when you try to scroll but have to wait for a prolonged period for some animation to complete.

artemiomorales replied:
Pardon, the comment here was out of date. The lightbox no longer closes on scroll; users needs to click on the overlay or close button to close it.

That being said, I was testing, and we *could* just get rid of the delay. It's less than half a second though, and if we did get rid of it, the lightbox in that case may look sloppy, especially to less technically inclined users.

While we *could* add robust handling and logic to optimize the UX for a canceled zoom animation, I'm not sure that would be worth the time invested. I feel it requires a lot of effort to close the lightbox then immediately scroll and beat the 450 millisecond delay, after which scroll input is recognized immediately anyway.

The robust handling could be added in the future, though I think that can be a separate PR.

dmsnell reviewed (commented):
there's talk in the patch about differentiating mobile and desktop behavior, but I didn't see anything that detects mobile or behaves differently on desktop other than the `isTouching` event.

is there a point where we could know that we're not on mobile and pre-emptively unload the scroll handler?

it's too bad that the extra runtime overhead is only meant for mobile where it would matter more.

code wise nothing jumped out other than some places where I think we might be able to communicate more clearly what we're doing.

michalczaplinski reviewed (approved):
Tested on a physical mobile device thanks to the [Gutenberg PR Previewer](https://playground.wordpress.net/gutenberg.html)

Looks great to me. Thanks Artemio! 👍

artemiomorales committed:
Rename attribute for clarity

artemiomorales committed:
Update comment

artemiomorales committed:
Update comments

artemiomorales commented:
@dmsnell Thanks for taking a look!

>there's talk in the patch about differentiating mobile and desktop behavior, but I didn't see anything that detects mobile or behaves differently on desktop other than the isTouching event.
>
>is there a point where we could know that we're not on mobile and pre-emptively unload the scroll handler?

I looked further into this and we actually need to use both listeners, on mobile and desktop, the reason being that many devices support both touch and mouse input. The Surface Pro laptop is one example; one can even use a mouse with an iPad or an iPhone. 

In these cases, we'll want to have the scroll callback, but still recognize the moment when touch input starts so that we can override the scroll reset and allow  pinch to zoom to work as expected.

artemiomorales committed:
Fix spacing

mikachan commented:
I just cherry-picked this PR to the commits-for-6.4-beta2 branch to get it included in the next release: 0d638036f3

getdave commented:
✅ I updated the [PHP Sync Tracking Issue for WP 6.5](https://github.com/WordPress/gutenberg/issues/57959) to note this PR does not require a backport for WP 6.5 as it seems to have been cherry picked into WP 6.4.
